File: SDKs/javascript/frameworks/app_kit/scroller.js
/* 
 * scroller.js
 * vienna
 * 
 * Created by Adam Beynon.
 * Copyright 2009 Adam Beynon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
 
include('app_kit/view');

// NSScroller Vertical
resource('NSScrollerBottomArrowNormal.png');    // 15 x 18
resource('NSScrollerTopArrowNormal.png');       // 15 x 30
resource('NSScrollerTopSlotNormal.png');        // 15 x 18
resource('NSScrollerVBackgroundNormal.png');    // 15 x 6

resource('NSScrollerTopKnobNormal.png');        // 15 x 10
resource('NSScrollerBottomKnobNormal.png');     // 15 x 10
resource('NSScrollerVerticalKnobNormal.png');   // 15 x 1

// NSScrollArrowPosition
var NSScrollerArrowsDefaultSetting	= 0;
var NSScrollerArrowsNone	       	= 2;

// NSUsableScrollerParts
var NSNoScrollerParts			    = 0;
var NSOnlyScrollerArrows		    = 1;
var NSAllScrollerParts			    = 2;

// NSScrollerPart
var NSScrollerNoPart			    = 0;
var NSScrollerDecrementPage		    = 1;
var NSScrollerKnob			        = 2;
var NSScrollerIncrementPage		    = 3;
var NSScrollerDecrementLine    	    = 4;
var NSScrollerIncrementLine	 	    = 5;
var NSScrollerKnobSlot			    = 6;

// NSScrollerArrow
var NSScrollerIncrementArrow        = 0;
var NSScrollerDecrementArrow	    = 1;

var NSScroller = NSView.extend({
    
    _isVertical: null,
    
    /**
        Proportion of the available area that the knob takes up
        
        0.0 being the min - this does not mean a size of 0.0. If below a min size
        (say 20px for example), then the knob should remian a min size, to ensure
        it is still visible and still scrollable
        
        1.0 being the max - it fills the slot 100%.
    */
    _knobProportion: null,
    
    /**
        value , between 0,0 and 1.0. the position depends upon the size of the
        knob. 
    */
    _value: null,
    
    /*
        @ivar _knobTrackStartPoint - NSPoint
        Used to maintain the start point that the mouse was pressed when tracking.
        It is important to know at what point the mouse was pressed so that the
        position within the knob can be calculated.
    */
    _knobTrackStartPoint: null,
    
    initWithCoder: function(aCoder) {
        this._super(aCoder);
        
        this._isVertical = (this._bounds.size.width < this._bounds.size.height) ? true : false;
        
        this._value = aCoder.decodeDoubleForKey("NSCurValue");
        if (!this._value)
            this._value = 1;
        
        this._knobProportion = aCoder.decodeDoubleForKey("NSPercent");
        if (!this._knobProportion)
            this._knobProportion = 1;
                    
        return this;
    },
    
    drawRect: function(aRect) {
        var c = NSGraphicsContext.currentContext().graphicsPort();
        
        if (this._isVertical) {
            // var slotRect = this.rectForPart(NSScrollerKnobSlot);
            this.drawKnobSlotInRect(this.rectForPart(NSScrollerKnobSlot), false);
            NSImage.imageNamed('NSScrollerTopSlotNormal.png').drawInRect(NSMakeRect(0, 0, 15, 18));
            
            this.drawArrow(NSScrollerIncrementArrow, false);
            this.drawArrow(NSScrollerDecrementArrow, false);
            
            // draw the non part (background)
            // NSImage.imageNamed('NSScrollerVBackgroundNormal.png').drawInRect(this.rectForPart(NSScrollerNoPart));
            
            this.drawKnob();
        }
        
    },
    
    drawParts: function() {
        
    },
    
    /**
        One value from NSScrollerPart
    */
    rectForPart: function(partCode) {
        if (this._isVertical) {
            switch (partCode) {
                case NSScrollerKnobSlot:
                    return NSMakeRect(0, 8, 15, this._bounds.size.height - 45);
                    break;
                case NSScrollerIncrementLine:
                    // bottom arrow (facing down)
                    return NSMakeRect(0, this._bounds.size.height - 18, 15, 18);
                    break;
                case NSScrollerDecrementLine:
                    // top arrow
                    return NSMakeRect(0, this._bounds.size.height - (18 + 30), 15, 30);
                    break;
                case NSScrollerNoPart:
                    // returns the area between slot and top arrow
                    return NSMakeRect(0, 18, 15, this._bounds.size.height - (18 + 30 + 18));
                    break;
                
                case NSScrollerKnob:
                console.log(this._value + ' ...... ' + this._knobProportion);
                    var slotRect = this.rectForPart(NSScrollerKnobSlot);
                    var scrollerHeight = this._knobProportion * slotRect.size.height;
                    var yOffset = ((slotRect.size.height - scrollerHeight) * this._value); // ((this._value * scrollerHeight) / 2);
                    return NSMakeRect(0, yOffset + slotRect.origin.y, 15, scrollerHeight);
                    break;
            }
        }
        return NSMakeRect(0, 0, 0, 0);
    },
    
    checkSpaceForParts: function() {
        
    },
    
    /*
        One from NSUsableScrollerParts
    */
    usableParts: function() {
        
    },
    
    /*
        One from NSScrollArrowPosition
    */
    setArrowsPosition: function(position) {
        
    },
    
    arrowsPosition: function() {
        
    },
    
    /*
        @param anArrow - NSScrollArrow
        @param highlight - Boolean
    */
    drawArrow: function(anArrow, highlight) {
        if (anArrow == NSScrollerIncrementArrow) {
            var theRect = this.rectForPart(NSScrollerIncrementLine);
            NSImage.imageNamed('NSScrollerBottomArrowNormal.png').drawInRect(theRect);
        }
        else if (anArrow == NSScrollerDecrementArrow) {
            var theRect = this.rectForPart(NSScrollerDecrementLine);
            NSImage.imageNamed('NSScrollerTopArrowNormal.png').drawInRect(theRect);
        }
        
    },
    
    drawKnob: function() {
        var knobRect = this.rectForPart(NSScrollerKnob);
        NSImage.imageNamed('NSScrollerTopKnobNormal.png').drawInRect(NSMakeRect(knobRect.origin.x, knobRect.origin.y, 15, 10));
        NSImage.imageNamed('NSScrollerVerticalKnobNormal.png').drawInRect(NSMakeRect(knobRect.origin.x, knobRect.origin.y + 10, 15, knobRect.size.height - 20));
        NSImage.imageNamed('NSScrollerBottomKnobNormal.png').drawInRect(NSMakeRect(knobRect.origin.x, knobRect.origin.y + 10 + (knobRect.size.height - 20), 15, 10));
    },
    
    /*
        @param slotRect - NSRect
        @param highlight - Boolean
    */
    drawKnobSlotInRect: function(slotRect, highlight) {
        NSImage.imageNamed('NSScrollerVBackgroundNormal.png').drawInRect(slotRect);
    },
    
    /*
        @param highlight - Boolean
    */
    highlight: function(flag) {
        
    },
    
    /*
        @param thePoint - NSPoint
        @return NSScrollerPart
    */
    testPart: function(thePoint) {
        if (NSPointInRect(thePoint, this.rectForPart(NSScrollerKnob)))
            return NSScrollerKnob;        
        return -1;
    },
    
    /*
        @param theEvent - NSEvent
    */
    trackKnob: function(theEvent) {
        var location = this.convertPointFromView(theEvent.locationInWindow(), null);
        // Temp fix for inverted co-ord (cocoa origin bottom left)
        location.y = this._bounds.size.height - location.y;
        this._knobTrackStartPoint = location;
        
        NSApplication.sharedApplication().bindEventsMatchingMask((NSLeftMouseUpMask | NSMouseMovedMask), this, function(theEvent) {
            
            if (theEvent.type() == NSLeftMouseUp) {
                NSApplication.sharedApplication().unbindEvents();
                return;
            }
            
            var location = this.convertPointFromView(theEvent.locationInWindow(), null);
            // Temp fix for inverted co-ord (cocoa origin bottom left)
            location.y = this._bounds.size.height - location.y;
            
            console.log(location.y - this._knobTrackStartPoint.y);
        });
    },
    
    /*
        @param theEvent - NSEvent
    */
    trackScrollButtons: function(theEvent) {
        
    },
    
    /*
        @return NSScrollerPart
    */
    hitPart: function() {
        
    },
    
    /*
        @return CGFloat
    */
    knobProportion: function() {
        return this._knobProportion;
    },
    
    /*
        @param knobProportion - CGFloat
    */
    setKnobProportion: function(knobProportion) {
        this._knobProportion = knobProportion;
        this.setNeedsDisplay(true);
    },
    
    /*
        @param theEvent - NSEvent
    */
    mouseDown: function(theEvent) {
        var location = this.convertPointFromView(theEvent.locationInWindow(), null);
        // Temp fix for inverted co-ord (cocoa origin bottom left)
        location.y = this._bounds.size.height - location.y;
        var theTarget = this.testPart(location);
        
        switch (theTarget) {
            case NSScrollerKnob:
                this.trackKnob(theEvent);
                break;
            default:
                break;
        }
    }
});

/*
    Scroller width. To keep in with Interface builder, scrollers are 15px
    wide (or height if horizontal scroller).
*/
NSScroller.scrollerWidth = function() {
    return 15;
};

File: frameworks/app_kit/lib/table_view/scroller.rb
# 
# scroller.rb
# vienna
# 
# Created by Adam Beynon.
# Copyright 2009 Adam Beynon.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

module Vienna
  
  SCROLLER_PARTS = {
    :none             => 0,
    :increment_line   => 1,
    :decrement_line   => 2,
    :increment_page   => 3,
    :decrement_page   => 4,
    :knob             => 5,
    :knob_slot        => 6
  }
  
  SCROLLER_ARROWS = {
    :increment_arrow  => 0,
    :decrement_arrow  => 1
  }
  
  SCROLL_ARROW_POSITIONS = {
    :none             => 1,
    :min_end          => 2,
    :max_end          => 3
  }
  
  USABLE_SCROLLER_PARTS = {
    :none             => 0,
    :all              => 1,
    :only_arrows      => 2
  }
  
  class Scroller < Control
    
    # V_KNOB_IMAGE = ThreePartImage.new([Image.image_named('scroller_vertical_knob_top'),
    #                                    Image.image_named('scroller_vertical_knob_middle'),
    #                                    Image.image_named('scroller_vertical_knob_bottom')], true)
    # 
    # H_KNOB_IMAGE = ThreePartImage.new([Image.image_named('scroller_horizontal_knob_left'),
    #                                    Image.image_named('scroller_horizontal_knob_middle',
    #                                    Image.image_named('scroller_horizontal_knob_right'))])
    
    def self.scroller_width
      17
    end
    
    def self.scroller_width_for_control_size(control_size)
      17
    end
    
    def initialize(frame)
      super frame
      @value = 0.0
      @knob_proportion = 1
    end
    
    def class_name
      @frame.width < @frame.height ? 'vn-vertical-scroller' : 'vn-horizontal-scroller'
    end
    
    def render(context)
      if context.first_time?
        context << "<div class='dec-line'></div>"
        context << "<div class='inc-line'></div>"
        context << "<div class='knob'><div class='start'</div><div class='middle'></div><div class='end'></div></div>"
        context.first_time = false
      end
      context.class_name = class_name
      
      context.selector :knob do |knob|
        knob.frame = rect_for_part(:knob)
      end
      
    end
    
    def draw_parts
      
    end
    
    # for vertical, height of decrement/inc line
    # for horizontal, it will be the offset width
    DECREMENT_LINE_SIZE = 22
    
    def rect_for_part(part)
      result = Rect.new(0, 0, 0, 0)
      increment_line = Rect.new(0, 0, @bounds.width, @bounds.height)
      decrement_line = Rect.new(0, 0, @bounds.width, @bounds.height)
      knob = Rect.new(0, 0, @bounds.width, @bounds.height)
      knob_slot = Rect.new(0, 0, @bounds.width, @bounds.height)
      
      if vertical?
        decrement_line.height = DECREMENT_LINE_SIZE
        increment_line.height = DECREMENT_LINE_SIZE
        increment_line.y = @bounds.height - DECREMENT_LINE_SIZE
        
        knob_slot.height = @bounds.height - (2 * DECREMENT_LINE_SIZE)
        knob_slot.y = DECREMENT_LINE_SIZE
        
        
        
        knob.height = knob_slot.height * @knob_proportion
        knob.y = ((knob_slot.height - knob.height) * @value) + knob_slot.y
      else # horizontal...
        decrement_line.width = DECREMENT_LINE_SIZE
        increment_line.width = DECREMENT_LINE_SIZE
        increment_line.y = @bounds.width - DECREMENT_LINE_SIZE
        
        knob_slot.width = @bounds.width - (2 * DECREMENT_LINE_SIZE)
        knob_slot.x = DECREMENT_LINE_SIZE
        
        knob.width = knob_slot.width * @knob_proportion
        knob.x = ((knob_slot.width - knob.width) * @value) + knob_slot.x
      end
      
      
      
      return case part
      when :none
        result
      when :increment_line
        if vertical?
          
        else
        
        end
      when :decrement_line
      
      when :increment_page
      
      when :decrement_page
      
      when :knob
        knob
      when :knob_slot
        knob_slot
      end      
    end
    
    def check_space_for_parts
      
    end
    
    def usable_parts
      
    end
    
    def arrows_position=(position)
      @arrows_position = position
    end
    
    def arrows_position
      @arrows_position
    end
    
    def control_tint=(control_tint)
      @control_tint = control_tint
    end
    
    def control_tint
      @control_tint
    end
    
    def control_size=(control_size)
      @control_size = control_size
    end
    
    def control_size
      @control_size
    end
    
    def draw_arrow(which_arrow, highlight:flag)
      
    end
    
    def draw_knob
      
    end
    
    def draw_knob_slot_in_rect(slot_rect, highlight:flag)
      
    end
    
    def highlight(flag)
      
    end
    
    def test_part(the_point)
      
    end
    
    def mouse_down(the_event)
      return unless enabled?
      
      location = convert_point(the_event.location_in_window, from_view:nil)
      
      # case test_part(location)
      # when condition
      #   
      # end
      
      # puts "#{location.x}, #{location.y}"
      track_knob(the_event)
    end
    
    def track_knob(the_event)
      original_value = @value
      mouse_down_point = convert_point(the_event.location_in_window, from_view:nil)
      slot_rect = rect_for_part(:knob_slot)
      knob_rect = rect_for_part(:knob)
      
      size = vertical? ? slot_rect.height - knob_rect.height : slot_rect.width - knob_rect.width
      
      App.bind_events [:left_mouse_up, :left_mouse_dragged] do |the_event|
        
        if the_event.type == :left_mouse_up
          App.unbind_events
        else # left_mouse_dragged
          location = convert_point(the_event.location_in_window, from_view:nil)
          delta = vertical? ? location.y - mouse_down_point.y : location.x - mouse_down_point.x

          self.float_value = Math.min(Math.max(0, original_value + (delta / size)), 1)
          self.needs_display = true   
          send_action(@action, to:@target)     
        end
      end
    end
    
    def track_scroll_buttons(the_event)
      
    end
    
    def hit_part
      
    end
    
    def knob_proportion
      @knob_proportion
    end
    
    def knob_proportion=(proportion)
      @knob_proportion = proportion
      self.needs_display = true
    end
    
    def float_value=(a_float)
      @value = a_float
    end
    
    def float_value
      @value
    end
    
    def double_value
      @value
    end
    
    def double_value=(a_double)
      @value = a_double
    end
    
    def action=(an_action)
      @action = an_action
    end
    
    def target=(a_target)
      @target = a_target
    end
    
    def vertical?
      @frame.width < @frame.height ? true : false
    end
    
  end
end
